flash/data/data_module.py

```diff
@@ -0,0 +1,315 @@
+            or 0 for Darwin platform.
+    """
+
+    preprocess_cls = Preprocess
+    postprocess_cls = Postprocess
+
+    def __init__(
+        self,
+        train_ds: Optional[AutoDataset] = None,
+        valid_ds: Optional[AutoDataset] = None,
+        test_ds: Optional[AutoDataset] = None,
+        predict_ds: Optional[AutoDataset] = None,
+        batch_size: int = 1,
+        num_workers: Optional[int] = None,
+    ):
+        super().__init__()
+        self._train_ds = train_ds
+        self._valid_ds = valid_ds
+        self._test_ds = test_ds
+        self._predict_ds = predict_ds
+
+        if self._train_ds is not None:
+            self.train_dataloader = self._train_dataloader
+
+        if self._valid_ds is not None:
+            self.val_dataloader = self._val_dataloader
+
+        if self._test_ds is not None:
+            self.test_dataloader = self._test_dataloader
+
+        if self._predict_ds is not None:
+            self.predict_dataloader = self._predict_dataloader
+
+        self.batch_size = batch_size
+
+        # TODO: figure out best solution for setting num_workers
+        if num_workers is None:
+            if platform.system() == "Darwin":
+                num_workers = 0
+            else:
+                num_workers = os.cpu_count()
+        self.num_workers = num_workers
+
+        self._data_pipeline = None
+        self._preprocess = None
+        self._postprocess = None
+
+        # this may also trigger data preloading
+        self.set_running_stages()
+
+    @staticmethod
+    def get_dataset_attribute(dataset: torch.utils.data.Dataset, attr_name: str, default: Optional[Any] = None) -> Any:
+        if isinstance(dataset, Subset):
+            return getattr(dataset.dataset, attr_name, default)
+
+        return getattr(dataset, attr_name, default)
+
+    @staticmethod
+    def set_dataset_attribute(dataset: torch.utils.data.Dataset, attr_name: str, value: Any) -> None:
+        if isinstance(dataset, Subset):
+            dataset = dataset.dataset
+        setattr(dataset, attr_name, value)
+
+    def set_running_stages(self):
+        if self._train_ds is not None:
+            self.set_dataset_attribute(self._train_ds, 'running_stage', RunningStage.TRAINING)
+
+        if self._valid_ds is not None:
+            self.set_dataset_attribute(self._valid_ds, 'running_stage', RunningStage.VALIDATING)
+
+        if self._test_ds is not None:
+            self.set_dataset_attribute(self._test_ds, 'running_stage', RunningStage.TESTING)
+
+        if self._predict_ds is not None:
+            self.set_dataset_attribute(self._predict_ds, 'running_stage', RunningStage.PREDICTING)
+
+    def _resolve_collate_fn(self, dataset: Dataset, running_stage: RunningStage) -> Optional[Callable]:
+        if isinstance(dataset, AutoDataset):
+            return self.data_pipeline.worker_preprocessor(running_stage)
+
+    def _train_dataloader(self) -> DataLoader:
```

Comment:
this fine - however the `DataModule` class has also a confusing method `.train_dataloader()` that returns 
`ValueError: num_samples should be a positive integer value, but got num_samples=0`

Reply:
this refers to the comment below https://github.com/PyTorchLightning/lightning-flash/pull/141#discussion_r594529259
I'll open an issue with more details